test/utils/setup.js

```diff
@@ -1,6 +1,61 @@
+const formatUtil = require('format-util');
 const createDOM = require('./createDOM');
 
 process.browser = true;
 
 createDOM();
 require('./init');
+
+const mochaHooks = {
+  beforeEach: [],
+  afterEach: [],
+};
+
+function throwOnUnexpectedConsoleMessages(methodName, expectedMatcher) {
+  const unexpectedCalls = [];
+
+  function logUnexpectedConsoleCalls(format, ...args) {
+    const message = formatUtil(format, ...args);
+    // Safe stack so that test dev can track where the unexpected console message was created.
+    const { stack } = new Error();
+
+    unexpectedCalls.push([
+      // first line includes the (empty) error message
+      // i.e. Remove the `Error:` line
+      // second line is this frame
+      stack.split('\n').slice(2).join('\n'),
+      message,
+    ]);
+  }
+  // eslint-disable-next-line no-console
+  console[methodName] = logUnexpectedConsoleCalls;
```

Comment:
Something is off with the overall logic. If you run the tests in watch mode and log something in the console, the test won't fail. However, if you do the same outside the watch mode, it works correctly. We can reproduce with this diff and `yarn test:watch`, `yarn test:unit`. 

```diff
diff --git a/packages/material-ui-lab/src/Alert/Alert.test.js b/packages/material-ui-lab/src/Alert/Alert.test.js
index c4de7927fd..9a88e421c8 100644
--- a/packages/material-ui-lab/src/Alert/Alert.test.js
+++ b/packages/material-ui-lab/src/Alert/Alert.test.js
@@ -5,7 +5,7 @@ import describeConformance from '@material-ui/core/test-utils/describeConformanc
 import Alert from './Alert';
 import Paper from '@material-ui/core/Paper';

-describe('<Alert />', () => {
+describe.only('<Alert />', () => {
   const mount = createMount();
   let classes;

@@ -20,4 +20,8 @@ describe('<Alert />', () => {
     refInstanceof: window.HTMLDivElement,
     skip: ['componentProp'],
   }));
+
+  it('should work', () => {
+    console.error('oops');
+  });
 });
```

Do we need this `mochaHooks` logic in the first place?

Reply:
This is an issue with `mocha`: https://github.com/mochajs/mocha/issues/4347

Reply:
Upvoted, thanks for the reference.

Reply:
I'm not going to work on that issue since it's trivial to get working: explicitly expect an error or negate that expectation. A test without assertion is not a placeholder for "no error".

Reply:
Yeah, I think that we can wait for mocha to solve it :). It took me some time to understand what was going wrong between the CI in my local env.  Now, it's documented

Reply:
It kind of was documented in this PR: https://github.com/mui-org/material-ui/blob/45cb77ff02484e608467baded0b280fadd286467/test/README.md#L53-L55

Reply:
Should've included `yarn test:watch` but I never use it. Should've explicitly requested review of this file so that we're all on the same page.

Reply:
👌